crates/rfc5321/src/tls.rs

```diff
@@ -19,38 +80,55 @@ pub struct TlsOptions {
 }
 
 impl TlsOptions {
+    /// Produce a TlsConnector for this set of TlsOptions.
+    /// We need to employ a cache around the verifier as loading
+    /// the system certificate store can be a non-trivial operation
+    /// and not be something we want to do repeatedly in a hot code
+    /// path.  The cache does unfortunately complicate some of the
+    /// internals here.
     pub fn build_tls_connector(&self) -> TlsConnector {
-        let cipher_suites = if self.rustls_cipher_suites.is_empty() {
-            provider::DEFAULT_CIPHER_SUITES
-        } else {
-            &self.rustls_cipher_suites
+        let key = RustlsCacheKey {
+            insecure: self.insecure,
+            rustls_cipher_suites: self.rustls_cipher_suites.clone(),
         };
 
-        let provider = Arc::new(CryptoProvider {
-            cipher_suites: cipher_suites.to_vec(),
-            ..provider::default_provider()
-        });
-
-        let config = ClientConfig::builder_with_provider(provider.clone())
-            .with_protocol_versions(tokio_rustls::rustls::DEFAULT_VERSIONS)
-            .expect("inconsistent cipher-suite/versions selected");
-
-        let config = if self.insecure {
+        let config = if let Some(config) = key.get() {
```

Comment:
Nit: I'd do `return TlsConnector::from(config);` here so that you can dedent the entirety of the `else` branch.